### PR TITLE
Maya: Update frame range of instances when clicking on "Reset frame range"

### PR DESCRIPTION
**Pull request by friquette**
## Changelog Description
Update "start frame", "end frame", "handle start" and "handle end" extra attributes of animation instances when using the "Reset frame range" of OP in Maya.

### CODE REVIEW

**Comment by tokejepsen** [comment]
Interesting!

What is the use case for this?
Normally the frame range of the instances would be validated and repaired if they were out of sync from the asset.

**Comment by BigRoy** [comment]
I agree with @tokejepsen - this doesn't seem to be the correct intended behavior. 

Also, be aware that the logic will not work for all instances. There will be instances who do not have frame range attributes. There will also be instances that are time-dependent but don't expose their time settings on the instance (e.g. renderlayers).

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:2130]: This check seems to be redundant since the `cmds.ls` call already filters to only nodes with an `.id` attribute.

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:2137]: There will be instance nodes that are not time dependent and don't have the relevant attributes. For those this will error.

**Review by BigRoy — COMMENTED** [review]
As mentioned I suspect this PR won't get merged since it doesn't seem to be the intended behavior of the method. However, here are some purely code based notes.

**Comment by OlivierOsotimehin** [comment]
Hello, 
This is was a feature requested to add fluidity to the artist work. If a supervisor tell artist there is a change of duration of a shot : 
[101-150] > [101-180] 

They will use the **rest frame range**  to get the last frame range from production tracking. This will get change for the frame in / out / handle of the scene itself but not on the exported entities (animation, camera , etc ) 

If publish as is, the sanity check is going to complain that things are not right.

Since we are already doing something (**reset frame range**) regarding the time we could ensure that all the entities that are dependent of time in the scene  are changed to the correct known value.

Otherwise, reset frame range should not be there and all should be repaired/ fix from the sanity check.

**Comment by tokejepsen** [comment]
> If publish as is, the sanity check is going to complain that things are not right.

This validator [Validate Frame Range](https://github.com/ynput/OpenPype/blob/develop/openpype/hosts/maya/plugins/publish/validate_frame_range.py), should fail if the instance frame range is out of sync with the asset.

Does this not happen for your pipeline?

**Comment by OlivierOsotimehin** [comment]
It does happen ! but we could make it faster to publish scene by changing also the instance to what we know to be the correct value when we use the "reset frame range" in the openpype menu. 

That's for sure a "comfort" request, that lead to less complaints of artist and more adhesion of pipeline

**Review by tokejepsen — CHANGES_REQUESTED** [review]
@friquette this feature will need to be behind a setting

**Comment by OlivierOsotimehin** [comment]
Hello, 
Yes the artist tend to click reset frame range when they are told that the frame range have change (which happens quite often in commercials).

And for sure we don't optimise cache time for exemple, we aim for stability (when the character is not visible there may be shadow, etc etc ) so we export always the full range with a little more even for motion blur raison

**Comment by tokejepsen** [comment]
@BigRoy what you think of this?

This feature is behind a setting, so studios need to opt in now. I would also suggest to maybe widen the scope of this to no just `animation` instances, this could come as a `family` filter setting. We would need to be more explicit in the code though about which instances are updated and and maybe how since not all instances are made the same.
Lastly I think we should refactor the code for collecting the instances to `lib` so the collector and `reset frame range` share the same logic.

**Comment by BigRoy** [comment]
> @BigRoy what you think of this?
> 
> This feature is behind a setting, so studios need to opt in now. 

Looks great to me. We would have it disabled but I can see it being useful for some.

We might want to add a `instances=True` argument to the `reset_frame_range` function because the function is used in other areas of the code to, e.g. on resetting render settings / creating a first render instance (if resetting on create is enabled in project settings, which I believe is the default behavior). As such, that would now ALSO reset the instances in the scene - making it very confusing. Then the calls that don't want to reset instances to begin with use `instances=False`.

> I would also suggest to maybe widen the scope of this to no just animation instances, this could come as a family filter setting.  > We would need to be more explicit in the code though about which instances are updated and and maybe how since not all instances are made the same.

It currently actually not only applies to `animation` family instances, but to any instance that has the relevant frame based attributes, so it'd also do it for pointcaches, yeti cache frame ranges, etc.

The logic wouldn't work for resetting frame ranges of renderlayer instances, because those attributes aren't on the renderlayer instance but set in the scene with potential renderlayer overrides, etc.

> Lastly I think we should refactor the code for collecting the instances to lib so the collector and reset frame range share the same logic.

That would make sense. It's good to note however that #4388 will already do that since there you'd get the instances from the Create Context instead. So I'm wondering whether it's worth it now. This also means that this feature will need refactoring in that PR to continue to work.

**Comment by tokejepsen** [comment]
@friquette could you implement this?

- add keyword `instances=True` to `reset_frame_range` and make sure it does not run on creating render settings.
- widen the scope of to other instances that have frame ranges, but not `render` family.

**Comment by BigRoy** [comment]
> widen the scope of to other instances that have frame ranges, but not render family.

As far as I know it already works for ANY instance that has the frame attributes - it doesn't seem to check anything related to animation instances or alike.

**Comment by tokejepsen** [comment]
> As far as I know it already works for ANY instance that has the frame attributes - it doesn't seem to check anything related to animation instances or alike.

Yup, you are right.

File: openpype/settings/entities/schemas/projects_schema/schema_project_maya.json
Comment (mkolar) [inline openpype/settings/entities/schemas/projects_schema/schema_project_maya.json:158]:
```suggestion
            "label": "Update publishable instances on Reset Frame Range",
```

Animation very much suggests that it only works for that family, but actually it updates all publishable instances

File: openpype/settings/entities/schemas/projects_schema/schema_project_maya.json
Comment (mkolar) [inline openpype/settings/entities/schemas/projects_schema/schema_project_maya.json:163]:
```suggestion
                    "label": "If enabled, the frame range and the handles of all the publishable instances will be updated when using the 'Reset Frame Range' functionality"
```

File: openpype/settings/entities/schemas/projects_schema/schema_project_maya.json
Comment (mkolar) [inline openpype/settings/entities/schemas/projects_schema/schema_project_maya.json:157]: This is not explicit enough in maya context. It should be clearer what it relates too

```suggestion
            "key": "update_publishable_frame_range",
```

File: openpype/settings/defaults/project_settings/maya.json
Comment (mkolar) [inline openpype/settings/defaults/project_settings/maya.json:50]:
```suggestion
    "update_publishable_frame_range": {
```

**Review by mkolar — CHANGES_REQUESTED** [review]
In general this makes pretty good sense in certain workflows. I'm not sure it actually add a whole lot of value compared to keeping identical functionality in the validators repair, but I have no major problem with it.

**Review by tokejepsen — CHANGES_REQUESTED** [review]
Add keyword `instances=True `to `reset_frame_range` and make sure it does not run on creating render settings.

**Comment by tokejepsen** [comment]
@friquette is this ready to test?

**Comment by friquette** [comment]
> @friquette is this ready to test?

Yes it is

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:2203]:
```suggestion
        instances (bool, Optional): Whether to update publishable instances.
            Defaults to True. Even when True this will only reset instances if
            `update_publishable_frame_range` is also enabled in project settings.
```

Should we clarify that even when True it only does so if also enabled in settings?

**Review by tokejepsen — CHANGES_REQUESTED** [review]
Getting this error when opening the settings:
```
Traceback (most recent call last):

  File "C:\Users\tokejepsen\OpenPype\openpype\tools\settings\settings\categories.py", line 564, in reset
    self._create_root_entity()

  File "C:\Users\tokejepsen\OpenPype\openpype\tools\settings\settings\categories.py", line 979, in _create_root_entity
    entity = ProjectSettings(

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\root_entities.py", line 654, in __init__
    super(ProjectSettings, self).__init__(schema_hub, reset)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\root_entities.py", line 76, in __init__
    self._item_initialization()

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\root_entities.py", line 202, in _item_initialization
    self._add_children(self.schema_data)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\root_entities.py", line 172, in _add_children
    child_obj = self.create_schema_object(children_schema, self)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\root_entities.py", line 231, in create_schema_object
    return self.schema_hub.create_schema_object(

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\lib.py", line 358, in create_schema_object
    return klass(schema_data, *args, **kwargs)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\base_entity.py", line 916, in __init__
    self._item_initialization()

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\dict_immutable_keys_entity.py", line 196, in _item_initialization
    self._add_children(self.schema_data)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\dict_immutable_keys_entity.py", line 163, in _add_children
    child_obj = self.create_schema_object(children_schema, self)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\base_entity.py", line 962, in create_schema_object
    return self.schema_hub.create_schema_object(*args, **kwargs)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\lib.py", line 358, in create_schema_object
    return klass(schema_data, *args, **kwargs)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\base_entity.py", line 916, in __init__
    self._item_initialization()

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\dict_immutable_keys_entity.py", line 196, in _item_initialization
    self._add_children(self.schema_data)

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\dict_immutable_keys_entity.py", line 146, in _add_children
    children_schemas = self.schema_hub.resolve_schema_data(

  File "C:\Users\tokejepsen\OpenPype\openpype\settings\entities\lib.py", line 308, in resolve_schema_data
    schema_type = schema_data["type"]

KeyError: 'type'
```

**Comment by BigRoy** [comment]
Just noticed that there's actually logic implemented that is VERY CLOSE to this PR [here](https://github.com/ynput/OpenPype/blob/281f6186e44d476a45b52d1c2292ee85b2a93463/openpype/hosts/maya/api/lib.py#L3069-L3093) but that's only triggered on task changed. We could technically merge/clean up that functionality and also use it in reset frame range.

Also, according to the reset frame range logic implemented here I think that might be setting the wrong frame ranges for some things if handles needed to be included? It's not setting start and end handles for example.

Anyway, might be good to merge/cleanup since the logic is pretty much a duplicate.

**Comment by tokejepsen** [comment]
> Just noticed that there's actually logic implemented that is VERY CLOSE to this PR [here](https://github.com/ynput/OpenPype/blob/281f6186e44d476a45b52d1c2292ee85b2a93463/openpype/hosts/maya/api/lib.py#L3069-L3093) but that's only triggered on task changed. We could technically merge/clean up that functionality and also use it in reset frame range.

Good spot! Yeah, lets merge the logic/code.

**Comment by tokejepsen** [comment]
@friquette could you have a look at @BigRoy feedback [here](https://github.com/ynput/OpenPype/pull/4793#issuecomment-1516771857), then we can merge this PR?

**Comment by friquette** [comment]
> @friquette could you have a look at @BigRoy feedback [here](https://github.com/ynput/OpenPype/pull/4793#issuecomment-1516771857), then we can merge this PR?

I don't really understand what I need to do here. Can you explain please ?

**Review by tokejepsen — CHANGES_REQUESTED** [review]
@friquette I still get [this error](https://github.com/ynput/OpenPype/pull/4793#pullrequestreview-1393401921) when opening the settings.

**Comment by friquette** [comment]
> @friquette I still get [this error](https://github.com/ynput/OpenPype/pull/4793#pullrequestreview-1393401921) when opening the settings.

It's fixed. I didn't see that I removed the 'type' key from a maya setting...

File: openpype/hosts/maya/api/lib.py
Comment (tokejepsen) [inline openpype/hosts/maya/api/lib.py:3064]: Could we get doc string for this?

Comment (tokejepsen) [reply openpype/hosts/maya/api/lib.py:3064]: Also maybe `update_instances_frame_range` would be more accurate?

File: openpype/hosts/maya/api/lib.py
Comment (tokejepsen) [inline openpype/hosts/maya/api/lib.py:3091]: Might need an attribute query here for `.id`.

**Review by tokejepsen — CHANGES_REQUESTED** [review]
Tested in Maya 2023 successfully. Just some minor code feedback.

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:3087]: This is never true because it's missing a `getAttr`?

**Review by tokejepsen — APPROVED** [review]

**Comment by tokejepsen** [comment]
@BigRoy you happy with this PR?

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:3095]:
```suggestion
        id_attr = "{}.id".format(instance)
        if cmds.getAttr(id_attr) != "pyblish.avalon.instance":
            continue
            
        if (
            cmds.attributeQuery('family', node=instance, exists=True) and
            cmds.getAttr("{}.family".format(instance)) == "render"
        ):
            continue
```

Re-ordered to check the `.id` attribute value first since if it's true it's also very likely it has the family attribute instead of first checking the family attribute on nodes with an `.id` attribute that might not be a publish instance.

I wonder if this way of writing is more readable for the family attribute if statement - so I've refactored that too. I believe this would also only format the string if the attribute existed to begin with.

Thoughts?

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:3077]: With the function name refactored to `update_instances_frame_range` the `new_` suffix to this variable seems a bit odd. I feel like we could just refactor these variables to: `asset_name` and `asset_data`.

**Review by BigRoy — APPROVED** [review]
> you happy with this PR?

I am overjoyed. I like it!

I did have two minor cosmetic notes but they aren't critical.

**Review by tokejepsen — CHANGES_REQUESTED** [review]
Lets get @BigRoy feedback in before merging.

**Review by tokejepsen — APPROVED** [review]
Just some code cosmetic suggestions.

**Review by BigRoy — APPROVED** [review]
Agree with the last cosmetics comment by @tokejepsen but other than that solid work.

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:356]: Can you revert this change so that we stick with the right naming as per this PR https://github.com/ynput/OpenPype/pull/4978#issuecomment-1551109533 ?

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:2233]: This function is called `get_frame_range`. It should **not** adjust the scene state.

**Review by BigRoy — COMMENTED** [review]
This PR is showing me many changes which seem to revert code that was originally better. Did resolving some merge conflicts maybe ended up the bad way?

File: openpype/hosts/maya/api/lib.py
Comment (BigRoy) [inline openpype/hosts/maya/api/lib.py:4050]: I suspect these shouldn't be removed?

**Comment by BigRoy** [comment]
@tokejepsen Could you review again? There've been quite some commits with changing code.

**Comment by mkolar** [comment]
Considering this is now behind a settings with and opt-in for a studio and that it's re-using existing code with some cleanup, I'm happy to merge it, provided someone can confirm the functionality itself is as intended.

**Review by mkolar — APPROVED** [review]

**Comment by ClementHector** [comment]
Can anyone confirm that this works to be merged?

**Comment by mkolar** [comment]
> Can anyone confirm that this works to be merged?

I was actually ready to merge it, but tons of commits came after the last approval, so we need to try it again.

File: openpype/hosts/maya/api/lib.py
Comment (tokejepsen) [reply openpype/hosts/maya/api/lib.py:2272]: @friquette can we get this in?

**Review by tokejepsen — APPROVED** [review]
Minor docs request.

Tested successfully in Maya 2023.

**Comment by tokejepsen** [comment]
@BigRoy what do you think of the code?

**Comment by friquette** [comment]
New PR cleaner here: #5201

**Pull request closed without merging**
